### DISCUSSION TITLE
Why the selection:cleared event is inconsistent when an activeSelection or an object is removed?

### DISCUSSION THREAD

**Discussion by mdawar** [fabricjs/fabric.js#6742] [General]
Hi,

I have a question about the `selection:cleared` event please.

It seems that when a single object is removed using `canvas.remove(obj)` this event is fired with [`{ target }`](https://github.com/fabricjs/fabric.js/blob/master/src/canvas.class.js#L1074) and not [`{ deselected }`](https://github.com/fabricjs/fabric.js/blob/master/src/canvas.class.js#L1122) which is the case when deselecting objects.

I discovered this when I was adding buttons that are displayed for specific objects only when selected, and I was using the `selection:*` events for this, I have implemented a workaround but I wanted to ask you about it before creating a pull request or an issue because a change like this may be backwards incompatible and maybe it should be changed in the next major version.

Thank you.

**Comment by asturur** [comment]
Can you better specify what are doing with fabric ( what are you removing and how ) and what events do you see firing?

**Reply by mdawar** [reply]
Hi,

I have create a [fiddle](https://jsfiddle.net/fc4rxkLo/) to show what I'm talking about.

To reproduce:

#### Remove a single object
1. Add text or circle
2. Remove the object using the remove button
You'll get the cleared event with the removed object as [`target`](https://github.com/fabricjs/fabric.js/blob/master/src/canvas.class.js#L1074)

#### Remove multiple objects
1. Add multiple objects (text, circle)
2. Select all of them and remove them using the remove button
You'll get the removed objects as the [`deselected`](https://github.com/fabricjs/fabric.js/blob/master/src/canvas.class.js#L1122) array.

As for what I'm doing in my app, I'm listening for all of the `selection:*` events and checking the `target`'s type to display UI elements specific to this type, I noticed that when I remove a single element I was getting `target` and not `deselected`, so my code was still showing these UI elements even though the object was deleted, that's not a problem anymore I'm now checking for this case but I wanted to let you know about this inconsistency, `Canvas._onObjectRemoved` is returning an object different than `Canvas._fireSelectionEvents`.
